### PR TITLE
OpenVX NN extension module.

### PR DESCRIPTION
**Pull request by lcskrishna**
1. Openvx NN Extension user kernels (vx_nn)
2. Added inference generator module. (utils/inference_generator).

### CODE REVIEW

File: vx_nn/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    src/slice_layer.cpp
+    src/image_tensor_converter.cpp
+    src/tensor_image_converter.cpp
+    src/tensor_add.cpp
+    src/tensor_subtract.cpp
+    src/tensor_multiply.cpp
+    src/tensor_convert_depth_node.cpp
+)
+
+include_directories (../../amdovx-core/openvx/include src)
```
Comment (rgiduthuri) [inline vx_nn/CMakeLists.txt]: Use `../deps/amdovx-core/openvx/include` instead of `../../amdovx-core/openvx/include`

**Comment by rgiduthuri** [comment]
please move proto2bin.cpp and bin2proto.cpp out of this pull request as they are not relevant to this project..

**Comment by rgiduthuri** [comment]
also replace ${CMAKE_BINARY_DIR}/proto with ${CMAKE_CURRENT_BINARY_DIR}/proto in utils/inference_generator/CMakeLists.txt

File: vx_nn/src/batch_normalization_layer.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    data->bnBias = clCreateBuffer(context, CL_MEM_READ_WRITE, size, NULL, NULL);
+    data->bnScale = clCreateBuffer(context, CL_MEM_READ_WRITE, size, NULL, NULL);
+    if (!data->bnBias && !data->bnScale) {
+        return VX_FAILURE;
+    }
+    
+    //input and output memory.
+    ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_BUFFER_OPENCL, &data->input_mem, sizeof(data->input_mem)));
+    ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_BUFFER_OPENCL, &data->bnScale, sizeof(data->bnScale)));
+    ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_BUFFER_OPENCL, &data->bnBias, sizeof(data->bnBias)));
```
Comment (rgiduthuri) [inline vx_nn/src/batch_normalization_layer.cpp]: Something is not right here. Why bnBias and bnScale OpenCL buffers are being created here. 
The vxQueryTensor with VX_TENSOR_BUFFER_OPENCL will return OpenCL buffers for use by processBatchNormalizationLayer.

File: vx_nn/src/batch_normalization_layer.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+    return VX_SUCCESS;
+}
+
+static vx_status VX_CALLBACK uninitializeBatchNormalizationLayer(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    BatchNormLayerLocalData * data = NULL;
+    ERROR_CHECK_STATUS(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    clReleaseMemObject(data->bnBias);
+    clReleaseMemObject(data->bnScale);
+    if (data) {
```
Comment (rgiduthuri) [inline vx_nn/src/batch_normalization_layer.cpp]: Trying clReleaseMemObject with cl_mem object that's owned by the OpenVX framework can make the system unstable.

File: vx_nn/src/convolution_layer.cpp [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    ERROR_CHECK_STATUS(vxSetNodeAttribute(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+
+    return VX_SUCCESS;
+}
+
+static vx_status VX_CALLBACK uninitializeConvolutionLayer(vx_node node, const vx_reference *parameters, vx_uint32 num)
+{
+    ConvolutionLayerLocalData * data = NULL;
+    ERROR_CHECK_STATUS(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    clReleaseMemObject(data->workspace);
```
Comment (rgiduthuri) [inline vx_nn/src/convolution_layer.cpp]: missing error check

File: vx_nn/src/deconvolution_layer.cpp [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+    return VX_SUCCESS;
+
+}
+
+static vx_status VX_CALLBACK uninitializeDeconvolutionLayer(vx_node node, const vx_reference *parameters, vx_uint32 num)
+{
+    DeconvolutionLayerLocalData * data = NULL;
+    ERROR_CHECK_STATUS(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    clReleaseMemObject(data->workspace);
```
Comment (rgiduthuri) [inline vx_nn/src/deconvolution_layer.cpp]: missing error check

File: vx_nn/src/fully_connected_layer.cpp [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    //add to node attribute.
+    ERROR_CHECK_STATUS(vxSetNodeAttribute(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    return VX_SUCCESS;
+}
+
+static vx_status VX_CALLBACK uninitializeFullyConnectedLayer(vx_node node, const vx_reference *parameters, vx_uint32 num)
+{
+    FullyConnectedLayerLocalData * data = NULL;
+    ERROR_CHECK_STATUS(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    clReleaseMemObject(data->workspace);
```
Comment (rgiduthuri) [inline vx_nn/src/fully_connected_layer.cpp]: missing error check

File: vx_nn/src/normalization_layer.cpp [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    ERROR_CHECK_STATUS(vxSetNodeAttribute(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+
+    return VX_SUCCESS;
+}
+
+static vx_status VX_CALLBACK uninitializeNormalizationLayer(vx_node node, const vx_reference *parameters, vx_uint32 num)
+{
+    NormalizationLayerLocalData * data = NULL;
+    ERROR_CHECK_STATUS(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    clReleaseMemObject(data->workspace);
```
Comment (rgiduthuri) [inline vx_nn/src/normalization_layer.cpp]: missing error check

File: vx_nn/src/pooling_layer.cpp [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+    ERROR_CHECK_STATUS(vxSetNodeAttribute(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    return VX_SUCCESS;
+}
+
+static vx_status VX_CALLBACK uninitializePoolingLayer(vx_node node, const vx_reference *parameters, vx_uint32 num)
+{
+    PoolingLayerLocalData * data = NULL;
+    ERROR_CHECK_STATUS(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    clReleaseMemObject(data->pooling_workspace);
```
Comment (rgiduthuri) [inline vx_nn/src/pooling_layer.cpp]: missing error check... please make sure that error checks are done throughout the code

File: vx_nn/src/batch_normalization_layer.cpp (hunk 2 above)
Comment (lcskrishna) [inline vx_nn/src/batch_normalization_layer.cpp]: Fixed the issue.

File: vx_nn/src/batch_normalization_layer.cpp (hunk 3 above)
Comment (lcskrishna) [inline vx_nn/src/batch_normalization_layer.cpp]: Fixed.

File: vx_nn/src/convolution_layer.cpp (hunk 4 above)
Comment (lcskrishna) [inline vx_nn/src/convolution_layer.cpp]: fixed.

File: vx_nn/src/deconvolution_layer.cpp (hunk 5 above)
Comment (lcskrishna) [inline vx_nn/src/deconvolution_layer.cpp]: fixed.

File: vx_nn/src/fully_connected_layer.cpp (hunk 6 above)
Comment (lcskrishna) [inline vx_nn/src/fully_connected_layer.cpp]: fixed.

File: vx_nn/src/normalization_layer.cpp (hunk 7 above)
Comment (lcskrishna) [inline vx_nn/src/normalization_layer.cpp]: fixed.

File: vx_nn/src/pooling_layer.cpp (hunk 8 above)
Comment (lcskrishna) [inline vx_nn/src/pooling_layer.cpp]: fixed.

**Pull request merged**